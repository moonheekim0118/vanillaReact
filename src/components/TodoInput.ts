import { debounce, throttling } from "../util/util";

interface Props {
    keyword: string;
    onInput: (any) => void;
    onSubmit: (any) => void;
}

const TodoInput = ({ keyword, onInput, onSubmit }: Props): Element => {
    const container = document.createElement("form");
    container.className = "Input-Container";

    function render() {
        const Input = document.createElement("input");
        Input.className = "Todo-Input";
        Input.value = keyword;

        const SubmitButton = document.createElement("button");
        SubmitButton.className = "Todo-Submit-Btn";
        SubmitButton.innerText = "🐥";
        SubmitButton.type = "submit";

        container.appendChild(Input);
        container.appendChild(SubmitButton);

        // todo 입력시 input 이벤트로 저장 (디바운싱 적용)
        container.addEventListener("input", debounce(onInput, 900));

        // todo 등록시 저장 (스로틀링 적용)
        container.addEventListener("submit", throttling(onSubmit, 900));

        return container;
    }

    return render();
};

export default TodoInput;
