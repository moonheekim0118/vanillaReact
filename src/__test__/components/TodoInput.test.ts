import "jsdom-global/register";
import TodoInput from "../../components/TodoInput";

describe("<TodoInput/>", () => {
    let container;
    let MockOnInput = jest.fn();
    let MockOnSubmit = jest.fn();

    let MockData = {
        keyword: "test",
        onInput: MockOnInput,
        onSubmit: MockOnSubmit
    };

    it("renders correctly", () => {
        container = TodoInput(MockData);
    });

    it("matches snapshot", () => {
        expect(container).toMatchSnapshot();
    });

    it("should call onInput if input happend", () => {
        container.dispatchEvent(new Event("input"));
        expect(MockOnInput).not.toHaveBeenCalled(); // 쓰로틀링 체크
        setTimeout(() => expect(MockOnInput).toHaveBeenCalled(), 900);
    });

    it("should call onSubmit if form is submited", () => {
        container.dispatchEvent(new Event("submit"));
        expect(MockOnSubmit).toHaveBeenCalled();
    });
});
