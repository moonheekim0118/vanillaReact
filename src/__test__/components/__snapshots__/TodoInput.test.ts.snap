// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoInput/> matches snapshot 1`] = `
<form
  class="Input-Container"
>
  <input
    class="Todo-Input"
  />
  <button
    class="Todo-Submit-Btn"
    type="submit"
  />
</form>
`;
